source/python/neuropod/tests/test_tensorflow_v2_packaging.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2021 UATC, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import tensorflow as tf
+import unittest
+from testpath.tempdir import TemporaryDirectory
+
+from neuropod.packagers import create_tensorflow_neuropod
+from neuropod.tests.utils import get_addition_model_spec, check_addition_model
+from neuropod.utils.eval_utils import RUN_NATIVE_TESTS
+
+
+def create_tf_addition_model():
+    """
+    A simple addition model
+    """
+
+    class Adder(tf.Module):
+        @tf.function(
+            input_signature=[
+                tf.TensorSpec(shape=None, dtype=tf.float32),
+                tf.TensorSpec(shape=None, dtype=tf.float32),
+            ]
+        )
+        def add(self, x, y):
+            return {"out": x + y}
+
+    return Adder()
+
+
+@unittest.skipIf(tf.__version__[0] != "2", "Skipping TF 2.x tests for TF 1.x")
+class TestTensorflowV2Packaging(unittest.TestCase):
+    def package_simple_addition_model(self, do_fail=False, **kwargs):
+        with TemporaryDirectory() as test_dir:
+            neuropod_path = os.path.join(test_dir, "test_neuropod")
+
+            # Get the input/output spec along with test data
+            kwargs.update(get_addition_model_spec(do_fail=do_fail))
+
+            # `create_tensorflow_neuropod` runs inference with the test data immediately
+            # after creating the neuropod. Raises a ValueError if the model output
+            # does not match the expected output.
+            create_tensorflow_neuropod(
```

Comment:
Can you add tests for use of saved_model? It seems like the logic added in this change isn't tested. check_addition_model will test the forward call right?

Reply:
> Can you add tests for use of saved_model? It seems like the logic added in this change isn't tested.

Passing in `trackable_obj` exercises the same code as passing in `saved_model_dir` at inference time.

The only line in the packaging code that isn't exercised (because we're not testing `saved_model_dir`) is 

https://github.com/uber/neuropod/blob/38091314e8544abf0b50403344a9b1b1f3fa4a6f/source/python/neuropod/backends/tensorflow/packager.py#L96-L97

(you can confirm by looking at code coverage if you'd like)

I'll add a test for it though :)

> check_addition_model will test the forward call right?

Yes. `create_tensorflow_neuropod` also runs inference (see the comment on the lines above this one):

```py
            # `create_tensorflow_neuropod` runs inference with the test data immediately
            # after creating the neuropod. Raises a ValueError if the model output
            # does not match the expected output.
```